### PR TITLE
Move SoftIRQ implementations to /kernel

### PR DESCRIPTION
**Pull request by cchanging**
To reduce TCB. The only concern is that this PR make `CpuLocalCell` public, which is easy to be misused. Yet its interfaces should be safe, so I think it's still okay.

### CODE REVIEW

**Comment by cchanging** [comment]
@boterinas rerun

File: kernel/src/lib.rs [hunk 1]
```diff
@@ -70,9 +70,8 @@ pub mod net;
 pub mod prelude;
 mod process;
 mod sched;
-pub mod softirq_id;
+pub mod softirq;
```
Comment (tatetian) [inline kernel/src/lib.rs:73]: Bottom-half mechanisms should be useful to all drivers, which may exists as separate crates. As such, bottom-half mechanism should not be extracted out of `ostd` into `kernel`, but as a component. So I think we should have a `softirq` crate.

File: ostd/src/trap/handler.rs [hunk 2]
```diff
@@ -1,7 +1,23 @@
...
+
+use spin::Once;
+
 use crate::{arch::irq::IRQ_LIST, cpu_local_cell, trap::TrapFrame};
 
+static BOTTOM_HALF_HANDLER: Once<Box<dyn Fn() + Sync + Send>> = Once::new();
+
+/// Register a function to the interrupt bottom half execution.
+///
+/// The function can only be registered once; subsequent calls will do nothing.
```
Comment (tatetian) [inline ostd/src/trap/handler.rs:13]: Explain more about it. What is so-called "bottom half"? When will the registered handler be called?

File: ostd/src/trap/handler.rs [hunk 3]
```diff
@@ -19,7 +35,11 @@ pub(crate) fn call_irq_callback_functions(trap_frame: &TrapFrame, irq_number: us
     crate::arch::interrupts_ack(irq_number);
 
     crate::arch::irq::enable_local();
-    crate::trap::softirq::process_pending();
+
+    // Process the execution in bottom half.
+    if let Some(handler) = BOTTOM_HALF_HANDLER.get() {
+        handler()
```
Comment (tatetian) [inline ostd/src/trap/handler.rs:41]: The user-given `handler` should be executed in the atomic context. Should call `disable_preempt` before entering the `handler`. The original `process_pending` includes a call to `disable_preempt`, which should be removed.

**Review by tatetian — COMMENTED** [review]
Good job! It is great to see softirq being removed from TCB.

**Review by tatetian — APPROVED** [review]
LGTM. Thanks for the contribution!

**Pull request merged**